File: arraysAndHashing/cppRealizations/twoSum.cpp
#include <map>
#include <vector>
#include <iostream>
using namespace std;

vector<int> twoSum(vector<int>& nums, int target) {
    map<int, int> hashMap;
    for(int i = 0; i < nums.size(); i++){
        int diff = target - nums[i];
        if (hashMap.find(diff) != hashMap.end()){
            return {hashMap[diff], i};
        } else {
            hashMap[nums[i]] = i;
        }
    }
    return {};
}

int main(){
    vector<int> nums = {2, 7, 11, 15};
    int target = 9;
    vector<int> result = twoSum(nums, target);
    for(int i = 0; i < result.size(); i++){
        cout << result[i] << endl;
    }
    return 0;
}

File: arraysAndHashing/twoSum.py
# bruteforce has a time complexity of O(n^2) and a space complexity of O(1)
# hashmap has a time complexity of O(n) and a space complexity of O(n)

class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        hashMap = {} # value : index
        for index, value in enumerate(nums):
            diff = target - value
            if diff in hashMap:
                return [hashMap[diff], index]
            else:
                hashMap[value] = index
        return {}